LUCENE-10154 NumericLeafComparator to define getPointValues

mayya-sharipova opened the pull request:
This patch adds getPointValues to NumericLeafComparatorsimilar how it
has getNumericDocValues.

Numeric Sort optimization with points relies on the assumption that
points and doc values record the same information, as we substitute
iterator over doc_values with one over points.

If we override getNumericDocValues it almost certainly means that whatever
PointValues NumericComparator is going to look at shouldn't be used to
skip non-competitive documents. Returning null for pointValues in this
case will force comparator NOT to use sort optimization with points,
and continue with a traditional way of iterating over doc values.

mayya-sharipova committed:
LUCENE-10154 NumericLeafComparator to define getPointValues

This patch adds getPointValues to NumericLeafComparatorsimilar how it
has getNumericDocValues.

Numeric Sort optimization with points relies on the assumption that
points and doc values record the same information, as we substitute
iterator over doc_values with one over points.

If we override getNumericDocValues it almost certainly means that whatever
PointValues NumericComparator is going to look at shouldn't be used to
skip non-competitive documents. Returning null for pointValues in this
case will force comparator NOT to use sort optimization with points,
and continue with a traditional way of iterating over doc values.

jpountz reviewed (commented):
I think there is also a NumericComparator in the lucene/expressions module that needs to be updated? And maybe we should update javadocs on the comparator constructor to clarify that the field name is used for the default implementation of getNumericDocValues/getPointsValues and that it is ok to pass a dummy value, e.g. when sorting by script, but then both methods should be overridden?

mayya-sharipova committed:
Address Adrien's feedback

mayya-sharipova commented:
@jpountz Thank you for your feedback. I've addressed your comments in c4814f248b976ae2ae23173fc27319cb69d1c17d.  This is ready for another round of review. 

> I think there is also a NumericComparator in the lucene/expressions module that needs to be updated?

Indeed, good comment. `ExpressionValueSource` extends `DoubleValuesSource`, `DoubleValuesSource` was updated to return `null` for points in c4814f248b976ae2ae23173fc27319cb69d1c17d.

jpountz reviewed (approved):
LGTM!

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into compartor-point-values

mayya-sharipova committed:
Add changes
